flocker/acceptance/endtoend/test_dataset.py

```diff
@@ -35,6 +40,114 @@ def test_dataset_creation(self, cluster):
+        unsupported={DatasetBackend.loopback},
+        reason="Does not maintain compute_instance_id across restarting "
+               "flocker (and didn't as of most recent release).")
+    @run_test_with(async_runner(timeout=timedelta(minutes=6)))
+    @require_cluster(1)
+    def test_upgrade(self, cluster):
+        """
+        Given a dataset created and used with the previously installable
+        version of flocker, uninstalling the previous version of flocker and
+        installing HEAD does not destroy the data on the dataset.
+        """
+        node = cluster.nodes[0]
+        SAMPLE_STR = '123456' * 100
+
+        upgrade_from_version = get_installable_version(version)
+
+        # Get the initial flocker version and setup a cleanup call to restore
+        # flocker to that version when the test is done.
+        d = cluster.client.version()
+        original_package_source = [None]
+
+        def setup_restore_original_flocker(version):
+            original_package_source[0] = (
+                self._get_package_source(
+                    default_version=str(version.get('flocker')) or None)
+            )
+            self.addCleanup(
+                lambda: cluster.install_flocker_version(
+                    original_package_source[0]))
+            return version
+
+        d.addCallback(setup_restore_original_flocker)
+
+        # Downgrade flocker to the most recent released version.
+        d.addCallback(
+            lambda _: cluster.install_flocker_version(
+                PackageSource(version=upgrade_from_version)))
+
+        # Create a dataset with the code from the most recent release.
+        d.addCallback(lambda _: create_dataset(self, cluster, node=node))
+        first_dataset = [None]
+
+        # Write some data to a file in the dataset.
+        def write_to_file(dataset):
+            first_dataset[0] = dataset
+            return node.run_as_root(
+                ['bash', '-c', 'echo "%s" > %s' % (
+                    SAMPLE_STR, os.path.join(dataset.path.path, 'test.txt'))])
+        d.addCallback(write_to_file)
+
+        # Upgrade flocker to the code under test.
+        d.addCallback(lambda _: cluster.install_flocker_version(
+            original_package_source[0]))
+
+        # Create a new dataset to convince ourselves that the new code is
+        # running.
+        d.addCallback(lambda _: create_dataset(self, cluster, node=node))
+
+        # Wait for the first dataset to be mounted again.
+        d.addCallback(lambda _: cluster.wait_for_dataset(first_dataset[0]))
+
+        # Verify that the file still has its contents.
+        def cat_and_verify_file(dataset):
+            output = []
+
+            def append_output(line):
+                output.append(line)
+            file_catting = node.run_as_root(
+                ['bash', '-c', 'cat %s' % (
+                    os.path.join(dataset.path.path, 'test.txt'))],
+                handle_stdout=append_output)
```

Comment:
Decided to do `lambda line: append_output(line)` instead.

...

Just kidding, obviously your suggestion was better and I'm using that.